File: src/algorithms/heaps/build-min-heap/sources/build-min-heap.py
# Build Min Heap — convert an arbitrary array into a valid min-heap in-place using sift-down
def build_min_heap(input_array):
    array = input_array[:]  # @step:initialize
    size = len(array)  # @step:initialize
    # Start from last non-leaf node and sift down each node toward root
    for start_idx in range(size // 2 - 1, -1, -1):  # @step:sift-down
        sift_down(array, start_idx, size)  # @step:sift-down
    return array  # @step:complete


def sift_down(array, start_idx, size):
    parent_idx = start_idx  # @step:sift-down
    while True:
        smallest_idx = parent_idx  # @step:sift-down
        left_idx = 2 * parent_idx + 1  # @step:sift-down
        right_idx = 2 * parent_idx + 2  # @step:sift-down
        # Find the smallest among parent, left child, and right child
        if left_idx < size and array[left_idx] < array[smallest_idx]:  # @step:sift-down
            smallest_idx = left_idx  # @step:sift-down
        if right_idx < size and array[right_idx] < array[smallest_idx]:  # @step:sift-down
            smallest_idx = right_idx  # @step:sift-down
        if smallest_idx == parent_idx:  # @step:sift-down
            break
        # Swap parent with the smallest child
        array[parent_idx], array[smallest_idx] = array[smallest_idx], array[parent_idx]  # @step:heap-swap
        parent_idx = smallest_idx  # @step:sift-down

File: src/algorithms/heaps/build-min-heap/sources/BuildMinHeap.java
// Build Min Heap — convert an arbitrary array into a valid min-heap in-place using sift-down
public class BuildMinHeap {
    public static int[] buildMinHeap(int[] inputArray) {
        int[] array = inputArray.clone(); // @step:initialize
        int size = array.length; // @step:initialize
        // Start from last non-leaf node and sift down each node toward root
        for (int startIdx = size / 2 - 1; startIdx >= 0; startIdx--) { // @step:sift-down
            siftDown(array, startIdx, size); // @step:sift-down
        }
        return array; // @step:complete
    }

    private static void siftDown(int[] array, int startIdx, int size) {
        int parentIdx = startIdx; // @step:sift-down
        while (true) {
            int smallestIdx = parentIdx; // @step:sift-down
            int leftIdx = 2 * parentIdx + 1; // @step:sift-down
            int rightIdx = 2 * parentIdx + 2; // @step:sift-down
            // Find the smallest among parent, left child, and right child
            if (leftIdx < size && array[leftIdx] < array[smallestIdx]) { // @step:sift-down
                smallestIdx = leftIdx; // @step:sift-down
            }
            if (rightIdx < size && array[rightIdx] < array[smallestIdx]) { // @step:sift-down
                smallestIdx = rightIdx; // @step:sift-down
            }
            if (smallestIdx == parentIdx) break; // @step:sift-down
            // Swap parent with the smallest child
            int temp = array[parentIdx]; // @step:heap-swap
            array[parentIdx] = array[smallestIdx]; // @step:heap-swap
            array[smallestIdx] = temp; // @step:heap-swap
            parentIdx = smallestIdx; // @step:sift-down
        }
    }
}
